network/framework/src/protocols/wire/messaging/v1/metadata.rs

```diff
@@ -0,0 +1,279 @@
+    pub fn application_send_time(&self) -> Option<SystemTime> {
+        self.application_send_time
+    }
+
+    /// Marks the message as having been fully sent over the network wire,
+    /// and emits the relevant latency metrics.
+    pub fn mark_message_as_sent(&mut self) {
+        // If this message is a streamed message fragment, there's no need to emit
+        // any metrics (we only emit metrics for the head and tail of streamed messages).
+        if self.message_stream_type == MessageStreamType::StreamedMessageFragment {
+            return;
+        }
+
+        // Otherwise, emit the latency metrics
+        if let Some(application_send_time) = self.application_send_time {
+            if let Some(wire_send_start_time) = self.wire_send_start_time {
+                // Calculate the application to wire send latency
+                let application_to_wire_latency = wire_send_start_time
+                    .duration_since(application_send_time)
+                    .unwrap_or_default()
+                    .as_secs_f64();
+
+                // Calculate the wire send latency
+                let wire_send_latency = wire_send_start_time
+                    .elapsed()
+                    .unwrap_or_default()
+                    .as_secs_f64();
+
+                // Update the application to wire latency metrics
+                counters::observe_message_send_latency(
+                    &counters::APTOS_NETWORK_MESSAGE_SEND_LATENCY,
+                    &self.network_id,
+                    &self.protocol_id,
+                    &self.message_send_type,
+                    &self.message_stream_type,
+                    MessageSendLatency::ApplicationToWire,
+                    application_to_wire_latency,
+                );
+
+                // Update the wire send latency metrics
+                counters::observe_message_send_latency(
+                    &counters::APTOS_NETWORK_MESSAGE_SEND_LATENCY,
+                    &self.network_id,
+                    &self.protocol_id,
+                    &self.message_send_type,
+                    &self.message_stream_type,
+                    MessageSendLatency::WireSend,
+                    wire_send_latency,
+                );
+            }
+        }
+    }
+
+    /// Returns a reference to the network ID
+    pub fn network_id(&self) -> &NetworkId {
+        &self.network_id
+    }
+
+    /// Returns a reference to the protocol ID
+    pub fn protocol_id(&self) -> &Option<ProtocolId> {
+        &self.protocol_id
+    }
+
+    /// Updates the message type
+    pub fn update_message_stream_type(&mut self, message_stream_type: MessageStreamType) {
+        self.message_stream_type = message_stream_type;
+    }
+
+    /// Updates the time at which the message started being sent over the wire
+    pub fn update_wire_send_start_time(&mut self) {
+        self.wire_send_start_time = Some(SystemTime::now());
```

Comment:
this should only be set for stream header right?

Reply:
It's set for all messages. If the message is streamed in chunks, then it's set for all chunks, too, i.e., head, middle and tail. When we emit the metrics (below), we only do it for head and tail chunks (so we can see how long it takes for the first send, and then how long the entire message took as all chunks passed through). 

Does this make sense?